Rusty error codes

Yoshanuikabundi opened the pull request:
This PR is a bit more involved. It refactors the errors in a number of ways that are a bit more involved than the previous PR. Changes are built around a new enum `ErrorCodes` that replicates the XDR error codes. This means that users of the library will never have to directly check an integer exposed by the C API, and it streamlines error processing within the crate as well. In addition, it paves the way for more granular and useful messages to the user. Errors are checked and a Result enum constructed in a single `ErrorCodes::check()` method, and then context is added with `Result::map()`. This means that several match statements can be replaced by a single line. The `Error` type is now a struct, similar to `std::io::Error`, which includes both the context of the error as well as the error code. The error code is provided through the `Error.source()` method, meaning that it plays nicely with the popular anyhow crate for pretty application error messages.

This is a bit more of a wild one and I'll totally understand if you don't like it :) I think the `ErrorCodes` enum is a dramatic improvement, and the other changes are mostly there to support it and improve the user experience when creating applications. The alternative design I considered was to refactor out the context/task that was being done when the error occurred, as the user has access to that themselves. So then `ErrorCodes` would be the only error type. But the dual error type approach might allow for better errors in the future.

I may also be overthinking things.

Yoshanuikabundi committed:
Broke out errors module and handle error codes as rust enum

Yoshanuikabundi committed:
Made error construction methods pub(crate)

Yoshanuikabundi committed:
Nicer error messages

Yoshanuikabundi committed:
Refactored ErrorCode as Error.source()

Yoshanuikabundi committed:
Streamlined error generation

Yoshanuikabundi committed:
Added documentation to the errors module

danijoo commented:
Hi.
good to see someone picks up my messy code and brings some structure to it ;)
I think it's a good idea to restructure errors like this.  Can you please rebase the branch on current master so this is mergeable (I don't want to `push -f` to your branch again :p

Yoshanuikabundi committed:
Added ToCString variant to Error.source()

Yoshanuikabundi committed:
Treat ToCString differently in Display based on whether it has a source

Yoshanuikabundi commented:
Haha I just hope it's good structure! Rebased and ready :) Your messy code is pretty great, I'm glad there's another Rust and GROMACS fan out there!

Yoshanuikabundi committed:
Corrected undefined behaviour when calling Trajectory.read() on improperly sized Frame

Yoshanuikabundi commented:
Was playing with iterators and had a segfault :/ calling Trajectory::read() with a frame of the wrong size passed a buffer of the wrong size to C, which lead to a segfault. I've fixed it by resizing the frame on every read(), which also required some changes to the error system. From the looks of things, Frame might've been the first module you wrote, coz its not quite as clean as the other modules - I'll have a look at Rusty-ising it tomorrow.

dnlbauer commented on src/lib.rs:
```diff
@@ -201,6 +201,10 @@ impl XTCTrajectory {
 impl Trajectory for XTCTrajectory {
     fn read(&mut self, frame: &mut Frame) -> Result<()> {
         let mut step: i32 = 0;
+        frame.resize(
```
I don't think the xtc or trr specification allows individual frames to differ in the number of atoms.
Checking this for every read could be a big performance loss when iterating over trajectories. I think we dont need this and users should be aware that it's their responsibility to have a correctly sized Frame (and a trajectory following the specification).

Yoshanuikabundi replied:
I think we need to do some check, coz we're claiming to be a safe wrapper and with no check we have undefined behaviour in safe code and that's not safe.

The resize should be very cheap --- get_num_atoms() is cached with lazy-init, so it's only a syscall the first time read() or get_num_atoms() is called on a trajectory. With the included iterator as written, later iterations will just compare the length of the frame vector to the cached u32 and then do nothing.

Yoshanuikabundi replied:
Are you worried about the frame.resize() call or the trajectory.get_num_atoms() call?

dnlbauer replied:
I actually forgot that `atom_num` is cached and not a call to the c_abi ;)

What do you think about raising an error if `get_num_atoms() != frame.len()`. I think this would be cleaner. If someone really passes a frame of incorrect size, there is very likely a bug in the piece of code that calls this function. Resizing just "hides" a bug in the users code.

I think checking is fine, but we should not alter an object that is passed to the function in unexpected ways (i.e. by resizing). From a user perspective, I would expect the function to raise an Error if I pass a wrongly sized Frame, not changing it.

Yoshanuikabundi replied:
I hear what you're saying, but the advantage of resizing for the user is that its more ergonomic I guess, and it follows expectations that the standard library sets. I can do a `Vec::new() `and be confident i'll get the right answer, and if I find later that I'm spending lots of time allocating I can change it to `Vec::with_capacity()`, and I think that expectation would be worth maintaining here.

But yeah, I'd be happy with an error, as long as `frame.len()` points to the actual size of the vector. At the moment it points to `Frame.num_atoms`, and `Frame.coords` and `Frame.num_atoms` are both pub, so a user could change one and get UB. I think the best way forward there would just be to remove the num_atoms field, and have len look directly at the length of coords. We'd also need to document that requirement because its either not in the docs or I missed it.

It's getting late here so I'll pick this up tomorrow :)

dnlbauer replied:
> I hear what you're saying, but the advantage of resizing for the user is that its more ergonomic I guess, and it follows expectations that the standard library sets. I can do a `Vec::new() `and be confident i'll get the right answer, and if I find later that I'm spending lots of time allocating I can change it to `Vec::with_capacity()`, and I think that expectation would be worth maintaining here.

But at this point, `with_capacity` has already been called with a wrong (too small) capacity. I think resizing would be equivalent to rust resizing vectors for you once you try to put something out of its bounds. i.e:

```Rust
let mut x = vec![0, 1, 2];
x[3] = 3;  // Does not resize x for you but errors.
```

Yoshanuikabundi replied:
Ok yep, you've convinced me. But I think we should change `with_capacity` to `with_len`, so there's no ambiguity 🙂

danijoo commented:
we can try to find a solution for the segfault issue in a separate issue/PR. I noticed it too but havent found a satisfying workaround..

Yoshanuikabundi committed:
Replaced frame resizing with a simple check

Yoshanuikabundi committed:
Cleaned up errors a bit

Yoshanuikabundi committed:
Reduced role of ErrorTask to simplify errors

Yoshanuikabundi committed:
Restricted ErrorTask to only being relevant to C API errors

Yoshanuikabundi committed:
Slightly clearer docs

Yoshanuikabundi committed:
Returned Error to being an enum, as other fields were unused

Yoshanuikabundi committed:
Clarified error message when read_num_atoms() fails during read()

Yoshanuikabundi committed:
Error::task() now checks sources

Yoshanuikabundi commented:
OK, I've made that change and reversed changes to the frame module. Read now checks that the frame is the right size, and there's a test (`test_wrong_size_frame`) that should trigger the UB if we regress. Fortuitously, having to support that change clarified what we actually require from the Error type, so I've streamlined and simplified it.

Commit:
code cleanup

Commit:
move check_code to lib

Commit:
tests

danijoo commented:
I made some small changes (mainly because of personal taste i guess):
- It does not feel "right" to me that a read/write/.. that does not result in an error has so much logic in `errors.rs`. Therefore, I moved `check_code` to `lib.rs` and also refractored this part a bit (It now returns an Option<Error>. This means there is some boilerplate in the methods (`if let Some(err)...`) but I feel like it makes the code easier to read and comprehend.
- The rest was just moving import statements to the top of files and stuff like that.

Let me know what you think. I'm pretty happy with the result and will merge if you agree. :+1:

Yoshanuikabundi committed:
Clearer errors if get_num_atoms fails during into_iter

Yoshanuikabundi committed:
Broke out next_inner

Yoshanuikabundi commented:
I think your changes are great :) I definitely have a bad habit of sticking to DRY even when it hurts readability.

I realised in the shower that our implementation of into_iter was broken in two ways - 
1. If we couldn't get the size of the frame from the trajectory, it'd generate an empty frame and then error when it tried to fill it, and that would be a wierd error - that's my bad for misunderstanding the semantics of frame and expecting it to resize when I removed an `unwrap` from `into_iter`
2. If the user kept a frame, and then resized it, and then called `next()`, the iterator would read the new size and fail

I've fixed both of these by adding a `num_atoms: Result<u32>` field to the iterator, which holds the result of the `get_num_atoms` call during iterator initialisation. This fixes both problems. It should be minimal impact, as `Error` is a very small type, the error and the number of atoms never have to be stored at the same time, and in the common case the check that happens on iteration is probably actually cheaper than dereferencing the old frame. I've also created inner functions for the iterators to reduce repetition and allow the inner next method to handle errors with `?`

This PR is pulling on a bit, sorry. I think I'm happy with it now. I only put these in this one rather than a new one because they are fundamentally about error handling.

Yoshanuikabundi committed:
Used cached result of self.trajectory.get_num_atoms() rather than storing it in TrajectoryIterator

danijoo commented:
Hi,
I benched it and the performance loss since 0d887fc515c68424ffe63c720acf0b6a0049da9e is negligible on a trajectory of relevant size (100 random atoms, 10,000 frames). Performance loss was  ~1%. I expected the boxing and cloning to have a bigger impact tbh, but the compiler probably optimizes this away:

test iterator::tests::bench_iter_old ... bench:  59,851,522 ns/iter (+/- 1,245,817)
test iterator::tests::bench_iter_old ... bench:  59,620,601 ns/iter (+/- 1,196,173)
test iterator::tests::bench_iter_old ... bench:  60,034,440 ns/iter (+/- 1,034,929)

test iterator::tests::bench_iter_new ... bench:  59,712,450 ns/iter (+/- 1,618,789)
test iterator::tests::bench_iter_new ... bench:  63,655,730 ns/iter (+/- 3,487,373)
test iterator::tests::bench_iter_new ... bench:  59,068,851 ns/iter (+/- 2,351,790)

Yoshanuikabundi commented:
Great! The box and clone only happens when there's an error, and I figured it was OK to be slow in that case :)

Commit:
reverted to single ReadNAtom error type

danijoo commented:
I reverted the Error types back to have a single ReadNAtoms error for both read and iterator. I dont think we really need two since its fundamentally the same error source.
